## Обязательная часть

Будем осуществлять работу с непростым [набором данных](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.csv) о состоянии здоровья лошадей, испытывающих кишечные колики. 

### Задание 1. Базовое изучение

Изучить представленный набор данных на основе [описания его столбцов](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.names) и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

### Задание 2. Работа с выбросами

В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

### Задание 3. Работа с пропусками

Рассчитать количество выбросов для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

In [1]:
import pandas as pd
import re

In [2]:
#В исходном датафрейме нет заголовков, я их получаю из horse_data.names
headers_list = []
with open('horse_data.names') as f:
    for line in f.readlines():
        if re.match(r'\d+:', line.strip()):
            headers_list.append(line.strip().replace(': ', '_').replace(' ', '_'))

for header in ['25_type_of_lesion', '26_type_of_lesion', '27_type_of_lesion']:
    headers_list.insert(-1, header)
headers_list[10] = '11_pain'

In [3]:
horses = pd.read_csv('horse_data.csv', names=headers_list, na_values='?')
horses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 28 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   1__surgery?                     299 non-null    float64
 1   2__Age                          300 non-null    int64  
 2   3__Hospital_Number              300 non-null    int64  
 3   4__rectal_temperature           240 non-null    float64
 4   5__pulse                        276 non-null    float64
 5   6__respiratory_rate             242 non-null    float64
 6   7__temperature_of_extremities   244 non-null    float64
 7   8__peripheral_pulse             231 non-null    float64
 8   9__mucous_membranes             253 non-null    float64
 9   10_capillary_refill_time        268 non-null    float64
 10  11_pain                         245 non-null    float64
 11  12_peristalsis                  256 non-null    float64
 12  13_abdominal_distension         244 

In [4]:
horses.head()

1__surgery?  2__Age  3__Hospital_Number  4__rectal_temperature  5__pulse  \
0          2.0       1              530101                   38.5      66.0   
1          1.0       1              534817                   39.2      88.0   
2          2.0       1              530334                   38.3      40.0   
3          1.0       9             5290409                   39.1     164.0   
4          2.0       1              530255                   37.3     104.0   

   6__respiratory_rate  7__temperature_of_extremities  8__peripheral_pulse  \
0                 28.0                            3.0                  3.0   
1                 20.0                            NaN                  NaN   
2                 24.0                            1.0                  1.0   
3                 84.0                            4.0                  1.0   
4                 35.0                            NaN                  NaN   

   9__mucous_membranes  10_capillary_refill_time  ...  19_packed_cell_volume  \
0                  NaN                       2.0  ...                   45.0   
1                  4.0                       1.0  ...                   50.0   
2                  3.0                       1.0  ...                   33.0   
3                  6.0                       2.0  ...                   48.0   
4                  6.0                       2.0  ...                   74.0   

   20_total_protein  21_abdominocentesis_appearance  \
0               8.4                             NaN   
1              85.0                             2.0   
2               6.7                             NaN   
3               7.2                             3.0   
4               7.4                             NaN   

   22_abdomcentesis_total_protein  23_outcome  24_surgical_lesion?  \
0                             NaN         2.0                    2   
1                             2.0         3.0                    2   
2                             NaN         1.0                    2   
3                             5.3         2.0                    1   
4                             NaN         2.0                    2   

   25_type_of_lesion  26_type_of_lesion  27_type_of_lesion  28_cp_data  
0              11300                  0                  0           2  
1               2208                  0                  0           2  
2                  0                  0                  0           1  
3               2208                  0                  0           1  
4               4300                  0                  0           2  

[5 rows x 28 columns]

In [5]:
horses.describe()

1__surgery?      2__Age  3__Hospital_Number  4__rectal_temperature  \
count   299.000000  300.000000        3.000000e+02             240.000000   
mean      1.397993    1.640000        1.085889e+06              38.167917   
std       0.490305    2.173972        1.529801e+06               0.732289   
min       1.000000    1.000000        5.184760e+05              35.400000   
25%       1.000000    1.000000        5.289040e+05              37.800000   
50%       1.000000    1.000000        5.303055e+05              38.200000   
75%       2.000000    1.000000        5.347275e+05              38.500000   
max       2.000000    9.000000        5.305629e+06              40.800000   

         5__pulse  6__respiratory_rate  7__temperature_of_extremities  \
count  276.000000           242.000000                     244.000000   
mean    71.913043            30.417355                       2.348361   
std     28.630557            17.642231                       1.045054   
min     30.000000             8.000000                       1.000000   
25%     48.000000            18.500000                       1.000000   
50%     64.000000            24.500000                       3.000000   
75%     88.000000            36.000000                       3.000000   
max    184.000000            96.000000                       4.000000   

       8__peripheral_pulse  9__mucous_membranes  10_capillary_refill_time  \
count           231.000000           253.000000                268.000000   
mean              2.017316             2.853755                  1.305970   
std               1.042428             1.620294                  0.477629   
min               1.000000             1.000000                  1.000000   
25%               1.000000             1.000000                  1.000000   
50%               2.000000             3.000000                  1.000000   
75%               3.000000             4.000000                  2.000000   
max               4.000000             6.000000                  3.000000   

       ...  19_packed_cell_volume  20_total_protein  \
count  ...             271.000000        267.000000   
mean   ...              46.295203         24.456929   
std    ...              10.419335         27.475009   
min    ...              23.000000          3.300000   
25%    ...              38.000000          6.500000   
50%    ...              45.000000          7.500000   
75%    ...              52.000000         57.000000   
max    ...              75.000000         89.000000   

       21_abdominocentesis_appearance  22_abdomcentesis_total_protein  \
count                      135.000000                      102.000000   
mean                         2.037037                        3.019608   
std                          0.804905                        1.968567   
min                          1.000000                        0.100000   
25%                          1.000000                        2.000000   
50%                          2.000000                        2.250000   
75%                          3.000000                        3.900000   
max                          3.000000                       10.100000   

       23_outcome  24_surgical_lesion?  25_type_of_lesion  26_type_of_lesion  \
count  299.000000           300.000000         300.000000         300.000000   
mean     1.551839             1.363333        3657.880000          90.226667   
std      0.737187             0.481763        5399.513513         649.569234   
min      1.000000             1.000000           0.000000           0.000000   
25%      1.000000             1.000000        2111.750000           0.000000   
50%      1.000000             1.000000        2673.500000           0.000000   
75%      2.000000             2.000000        3209.000000           0.000000   
max      3.000000             2.000000       41110.000000        7111.000000   

       27_type_of_lesion  28_cp_data  
count         300.000000  300.000000  
mean    

## Анализ столбцов 1 - 4
Мне показалось удобным поделить df на кусочки по 4 столбца и тупо последовательно посмотреть на каждый.

In [6]:
horses_four = horses[headers_list[:4]]
horses_four.head()

1__surgery?  2__Age  3__Hospital_Number  4__rectal_temperature
0          2.0       1              530101                   38.5
1          1.0       1              534817                   39.2
2          2.0       1              530334                   38.3
3          1.0       9             5290409                   39.1
4          2.0       1              530255                   37.3

Судя по описанию и по первым строчкам, в первых трех столбцах категориальные данные, в четвертом - количественные. План действий следующий:
* В столбцах 1 и 2, судя по описанию, возможны только 2 вида значений: 1 и 2. Соответственно, нужно просто проверить, в каком соотношении они встречаются, и нет ли ошибок (судя по тому, что в столбце Age уже видна девятка, ошибки есть).
* В столбце 3 интересно выяснить, много ли повторяющихся значений (есть ли лошади, которые лечились несколько раз)
* Для столбца 4 нужно рассчитать базовые метрики

In [7]:
horses_four.groupby(['1__surgery?', '2__Age']).count()

3__Hospital_Number  4__rectal_temperature
1__surgery? 2__Age                                           
1.0         1                      162                    125
            9                       18                     15
2.0         1                      113                     94
            9                        6                      5

In [8]:
# Смотрю, сколько в столбце 3 не уникальных ячеек
horses_four.groupby(['3__Hospital_Number']).count().groupby(['2__Age']).count()

1__surgery?  4__rectal_temperature
2__Age                                    
1               268                    268
2                16                     16

In [9]:
# Основные метрики для столбца 4
print(horses_four['4__rectal_temperature'].describe())
print(horses_four['4__rectal_temperature'].round(2). mode()[0])
print(horses_four['4__rectal_temperature'].median())
print(horses_four['4__rectal_temperature'].max() - horses_four['4__rectal_temperature'].min())

count    240.000000
mean      38.167917
std        0.732289
min       35.400000
25%       37.800000
50%       38.200000
75%       38.500000
max       40.800000
Name: 4__rectal_temperature, dtype: float64
38.0
38.2
5.399999999999999


In [10]:
# Выбросы для столбца 4
q1 = horses_four['4__rectal_temperature'].quantile(0.25)
q3 = horses_four['4__rectal_temperature'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr)
upper_bound = q3 + (1.5 * iqr)
without_outliers = horses_four[horses_four['4__rectal_temperature'].between(lower_bound, upper_bound)]
print(without_outliers['4__rectal_temperature'].describe())
print(without_outliers['4__rectal_temperature'].round(2). mode()[0])
print(without_outliers['4__rectal_temperature'].median())
print(without_outliers['4__rectal_temperature'].std())
print(without_outliers['4__rectal_temperature'].max() - without_outliers['4__rectal_temperature'].min())

count    226.000000
mean      38.169027
std        0.571949
min       36.800000
25%       37.800000
50%       38.200000
75%       38.500000
max       39.500000
Name: 4__rectal_temperature, dtype: float64
38.0
38.2
0.5719486606358535
2.700000000000003


In [11]:
pd.concat([horses_four, without_outliers]).drop_duplicates(keep=False).dropna(subset=['4__rectal_temperature'])

1__surgery?  2__Age  3__Hospital_Number  4__rectal_temperature
20           1.0       1              530157                   39.9
44           1.0       1              535407                   35.4
75           1.0       9              534092                   39.7
80           1.0       1              527518                   36.4
99           2.0       1              530002                   39.6
118          1.0       1              533983                   36.5
141          2.0       1              522979                   36.0
238          2.0       1              528702                   36.1
251          2.0       1              527940                   36.6
259          1.0       1              527734                   40.8
281          2.0       1              530624                   40.0
298          1.0       1              530612                   36.5

### Предварительные выводы по столбцам 1 - 4
1. Данные столбца surgery?, видимо, надо смотреть в сочетании с другими столбцами. Пока ничего определенного сказать нельзя.
2. В столбце Age вместо 2 (согласно описанию) почему-то используется 9. Скорее всего, это новое обозначение для молодых лошадей (их же не может вообще не быть, если они заявлены?), но полной уверенности у меня нет.
3. В столбце 3 найдено всего 16 повторяющихся значений, ни одна лошадь не наблюдалась больше двух раз. Соответственно, каждый случай можно рассматривать как уникальный.
4. Средние значения данных в столбце 4 составляют 38.0 - 38.2 градуса. Эти значения лишь незначительно меняются при удалении выбросов. Норма, согласно описанию, составляет 37.8 градусов. Таким образом, температура заболевших лошадей слегка повышена (это соответствует информации в описании).

## Анализ столбцов 5 - 8

In [12]:
# В каждую следующую группу я добавляю столбец 3, потому что в нем точно нет пропусков и удобно делать по нему группировку.
horses_eight = horses[[headers_list[2]] + headers_list[4:8]]
horses_eight.head()

3__Hospital_Number  5__pulse  6__respiratory_rate  \
0              530101      66.0                 28.0   
1              534817      88.0                 20.0   
2              530334      40.0                 24.0   
3             5290409     164.0                 84.0   
4              530255     104.0                 35.0   

   7__temperature_of_extremities  8__peripheral_pulse  
0                            3.0                  3.0  
1                            NaN                  NaN  
2                            1.0                  1.0  
3                            4.0                  1.0  
4                            NaN                  NaN

In [13]:
horses_eight.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 5 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   3__Hospital_Number             300 non-null    int64  
 1   5__pulse                       276 non-null    float64
 2   6__respiratory_rate            242 non-null    float64
 3   7__temperature_of_extremities  244 non-null    float64
 4   8__peripheral_pulse            231 non-null    float64
dtypes: float64(4), int64(1)
memory usage: 11.8 KB


### План действий:
* Для столбцов 5 и 6 рассчитать базовые метрики
* Для столбцов 7 и 8 сделать группировку данных, найти самое частое значение

In [14]:
# Основные метрики для столбца 5
print(horses_eight['5__pulse'].describe())
print(horses_eight['5__pulse'].round(2). mode()[0])
print(horses_eight['5__pulse'].median())
print(horses_eight['5__pulse'].max() - horses_eight['5__pulse'].min())

count    276.000000
mean      71.913043
std       28.630557
min       30.000000
25%       48.000000
50%       64.000000
75%       88.000000
max      184.000000
Name: 5__pulse, dtype: float64
48.0
64.0
154.0


In [15]:
# Выбросы для столбца 5
q1 = horses_eight['5__pulse'].quantile(0.25)
q3 = horses_eight['5__pulse'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr)
upper_bound = q3 + (1.5 * iqr)
without_outliers = horses_eight[horses_eight['5__pulse'].between(lower_bound, upper_bound)]
print(without_outliers['5__pulse'].describe())
print(without_outliers['5__pulse'].round(2). mode()[0])
print(without_outliers['5__pulse'].median())
print(without_outliers['5__pulse'].max() - without_outliers['5__pulse'].min())

count    271.000000
mean      70.258303
std       26.082484
min       30.000000
25%       48.000000
50%       64.000000
75%       88.000000
max      146.000000
Name: 5__pulse, dtype: float64
48.0
64.0
116.0


In [16]:
pd.concat([horses_eight, without_outliers]).drop_duplicates(keep=False).dropna(subset=['5__pulse'])

3__Hospital_Number  5__pulse  6__respiratory_rate  \
3               5290409     164.0                 84.0   
41              5288249     150.0                 72.0   
55              5282839     160.0                 20.0   
255             5294539     184.0                 84.0   
275             5297159     150.0                 50.0   

     7__temperature_of_extremities  8__peripheral_pulse  
3                              4.0                  1.0  
41                             NaN                  NaN  
55                             3.0                  NaN  
255                            1.0                  NaN  
275                            1.0                  3.0

In [17]:
# Основные метрики для столбца 6
print(horses_eight['6__respiratory_rate'].describe())
print(horses_eight['6__respiratory_rate'].round(2). mode()[0])
print(horses_eight['6__respiratory_rate'].median())
print(horses_eight['6__respiratory_rate'].max() - horses_eight['5__pulse'].min())

count    242.000000
mean      30.417355
std       17.642231
min        8.000000
25%       18.500000
50%       24.500000
75%       36.000000
max       96.000000
Name: 6__respiratory_rate, dtype: float64
20.0
24.5
66.0


In [18]:
# Выбросы для столбца 6
q1 = horses_eight['6__respiratory_rate'].quantile(0.25)
q3 = horses_eight['6__respiratory_rate'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr)
upper_bound = q3 + (1.5 * iqr)
without_outliers = horses_eight[horses_eight['6__respiratory_rate'].between(lower_bound, upper_bound)]
print(without_outliers['6__respiratory_rate'].describe())
print(without_outliers['6__respiratory_rate'].round(2). mode()[0])
print(without_outliers['6__respiratory_rate'].median())
print(without_outliers['6__respiratory_rate'].max() - without_outliers['6__respiratory_rate'].min())

count    225.000000
mean      26.715556
std       11.461464
min        8.000000
25%       18.000000
50%       24.000000
75%       35.000000
max       60.000000
Name: 6__respiratory_rate, dtype: float64
20.0
24.0
52.0


In [19]:
pd.concat([horses_eight, without_outliers]).drop_duplicates(keep=False).dropna(subset=['6__respiratory_rate'])

3__Hospital_Number  5__pulse  6__respiratory_rate  \
3               5290409     164.0                 84.0   
39              5277409     146.0                 96.0   
41              5288249     150.0                 72.0   
82              5290759     100.0                 80.0   
84               529849      60.0                 80.0   
103             5292489     140.0                 68.0   
106              528668      52.0                 96.0   
120              528682      54.0                 66.0   
125              528461      42.0                 68.0   
186             5278331      64.0                 90.0   
208              535338      88.0                 80.0   
229              534597     120.0                 70.0   
244             5291409     124.0                 88.0   
255             5294539     184.0                 84.0   
265              528590      56.0                 68.0   
269              527883      64.0                 90.0   
295              533886     120.0                 70.0   

     7__temperature_of_extremities  8__peripheral_pulse  
3                              4.0                  1.0  
39                             NaN                  NaN  
41                             NaN                  NaN  
82                             3.0                  1.0  
84                             1.0                  3.0  
103                            1.0                  1.0  
106                            NaN                  3.0  
120                            1.0                  1.0  
125                            4.0                  1.0  
186                            2.0                  3.0  
208                            3.0                  3.0  
229                            NaN                  NaN  
244                            1.0                  3.0  
255                            1.0                  NaN  
265                            3.0                  1.0  
269                            2.0                  3.0  
295                            4.0                  NaN

In [20]:
horses_eight.groupby(['7__temperature_of_extremities']).count()['3__Hospital_Number']

7__temperature_of_extremities
1.0     78
2.0     30
3.0    109
4.0     27
Name: 3__Hospital_Number, dtype: int64

In [21]:
horses_eight.groupby(['8__peripheral_pulse']).count()['3__Hospital_Number']

8__peripheral_pulse
1.0    115
2.0      5
3.0    103
4.0      8
Name: 3__Hospital_Number, dtype: int64

### Предварительные выводы по столбцам 5 - 8
1. В столбце 5 минимальное значение соответствует нижней границе нормы (30-40). В описании сказано, что у некоторых лошадей ("athletic horses") нормальное значение может быть ниже нормы. То есть в данной выборке есть тенденция к увеличению пульса. Выбросы есть только верхние, довольно значительные. В целом они выглядят логичными: можно предположить, что при коликах пульс лошади в целом увеличивается, а иногда увеличивается очень сильно.
2. Показатель "Частота дыхания" имеет в описании пометку "Полезность под сомнением из-за большого разброса данных". В целом картина та же, что для столбца 6: минимальное значение совпадает с нижней границей нормы (8). Средние значения значительно выше нормы (20-30 без выбросов). Выбросы идут только вверх. Так же возникает ощущение, что эти выбросы - естественные: разрыва в значениях нет.
3. Обычно отмечается нормальная или сниженная температура конечностей, нормальный или сниженный периферический пульс. Остальные значения встречаются значительно реже.

## Анализ столбцов 9 - 12

In [22]:
horses_twelve = horses[[headers_list[2]] + headers_list[8:12]]
horses_twelve.head()

3__Hospital_Number  9__mucous_membranes  10_capillary_refill_time  11_pain  \
0              530101                  NaN                       2.0      5.0   
1              534817                  4.0                       1.0      3.0   
2              530334                  3.0                       1.0      3.0   
3             5290409                  6.0                       2.0      2.0   
4              530255                  6.0                       2.0      NaN   

   12_peristalsis  
0             4.0  
1             4.0  
2             3.0  
3             4.0  
4             NaN

In [23]:
horses_twelve.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   3__Hospital_Number        300 non-null    int64  
 1   9__mucous_membranes       253 non-null    float64
 2   10_capillary_refill_time  268 non-null    float64
 3   11_pain                   245 non-null    float64
 4   12_peristalsis            256 non-null    float64
dtypes: float64(4), int64(1)
memory usage: 11.8 KB


В этом блоке все столбцы категориальные, с ограниченным набором значений. Поэтому просто сделаю группировки и посмотрю, какие значения встречаются чаще, а какие - реже.

In [24]:
horses_twelve.groupby(['9__mucous_membranes']).count()['3__Hospital_Number']

9__mucous_membranes
1.0    79
2.0    30
3.0    58
4.0    41
5.0    25
6.0    20
Name: 3__Hospital_Number, dtype: int64

In [25]:
horses_twelve.groupby(['10_capillary_refill_time']).count()['3__Hospital_Number']

10_capillary_refill_time
1.0    188
2.0     78
3.0      2
Name: 3__Hospital_Number, dtype: int64

In [26]:
horses_twelve.groupby(['11_pain']).count()['3__Hospital_Number']

11_pain
1.0    38
2.0    59
3.0    67
4.0    39
5.0    42
Name: 3__Hospital_Number, dtype: int64

In [27]:
horses_twelve.groupby(['12_peristalsis']).count()['3__Hospital_Number']

12_peristalsis
1.0     39
2.0     16
3.0    128
4.0     73
Name: 3__Hospital_Number, dtype: int64

### Предварительные выводы по столбцам 9 - 12
1. В столбцах 9 и 11 данные распределены достаточно равномерно. Лидирующие значения: в столбце 9 - 1 (слизистые оболочки нормального розового цвета), в столбце 11 - 3 (периодическая умеренная боль).
2. В столбце 10 (время наполнения капилляров) преобладает значение 1(быстрое, соответствует норме). Появляются два значения 3, не указанные в описании. Вероятно, это ошибка: граница между значениями 1 и 2 - 3 секунды, легко ошибиться и поставить цифру 3 вместо 2.
3. В столбце 12 (перистальтика) превалирует значение 3 (сниженная подвижность). В описании, насколько я смогла понять, сказано, что этот показатель характерен для колик.

## Анализ столбцов 13 - 16

In [28]:
horses_16 = horses[[headers_list[2]] + headers_list[12:16]]
horses_16.head()

3__Hospital_Number  13_abdominal_distension  14_nasogastric_tube  \
0              530101                      4.0                  NaN   
1              534817                      2.0                  NaN   
2              530334                      1.0                  NaN   
3             5290409                      4.0                  1.0   
4              530255                      NaN                  NaN   

   15_nasogastric_reflux  16_nasogastric_reflux_PH  
0                    NaN                       NaN  
1                    NaN                       NaN  
2                    NaN                       NaN  
3                    2.0                       5.0  
4                    NaN                       NaN

In [29]:
horses_16.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   3__Hospital_Number        300 non-null    int64  
 1   13_abdominal_distension   244 non-null    float64
 2   14_nasogastric_tube       196 non-null    float64
 3   15_nasogastric_reflux     194 non-null    float64
 4   16_nasogastric_reflux_PH  53 non-null     float64
dtypes: float64(4), int64(1)
memory usage: 11.8 KB


В этой группе числовые данные только в столбце 16. Это столбец с самым большим количеством пропусков, в нем всего 53 значения. Но я все-таки попытаюсь получить для него метрики.

In [30]:
horses_16.groupby(['13_abdominal_distension']).count()['3__Hospital_Number']

13_abdominal_distension
1.0    76
2.0    65
3.0    65
4.0    38
Name: 3__Hospital_Number, dtype: int64

In [31]:
horses_16.groupby(['14_nasogastric_tube']).count()['3__Hospital_Number']

14_nasogastric_tube
1.0     71
2.0    102
3.0     23
Name: 3__Hospital_Number, dtype: int64

In [32]:
horses_16.groupby(['15_nasogastric_reflux']).count()['3__Hospital_Number']

15_nasogastric_reflux
1.0    120
2.0     35
3.0     39
Name: 3__Hospital_Number, dtype: int64

In [33]:
horses_16.groupby(['16_nasogastric_reflux_PH']).count()['3__Hospital_Number']

16_nasogastric_reflux_PH
1.0    2
1.5    1
2.0    9
3.0    3
3.5    1
4.0    3
4.3    1
4.4    1
4.5    3
5.0    4
5.3    1
5.4    1
5.5    4
5.7    1
6.0    1
6.2    1
6.5    5
7.0    8
7.2    1
7.5    2
Name: 3__Hospital_Number, dtype: int64

In [34]:
# Основные метрики для столбца 16
print(horses_16['16_nasogastric_reflux_PH'].describe())
print(horses_16['16_nasogastric_reflux_PH'].round(2). mode()[0])
print(horses_16['16_nasogastric_reflux_PH'].median())
print(horses_16['16_nasogastric_reflux_PH'].max() - horses_16['16_nasogastric_reflux_PH'].min())

count    53.000000
mean      4.707547
std       1.982311
min       1.000000
25%       3.000000
50%       5.000000
75%       6.500000
max       7.500000
Name: 16_nasogastric_reflux_PH, dtype: float64
2.0
5.0
6.5


In [35]:
# Выбросы для столбца 16
q1 = horses_16['16_nasogastric_reflux_PH'].quantile(0.25)
q3 = horses_16['16_nasogastric_reflux_PH'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr)
upper_bound = q3 + (1.5 * iqr)
without_outliers = horses_16[horses_16['16_nasogastric_reflux_PH'].between(lower_bound, upper_bound)]
print(without_outliers['16_nasogastric_reflux_PH'].describe())
print(without_outliers['16_nasogastric_reflux_PH'].round(2). mode()[0])
print(without_outliers['16_nasogastric_reflux_PH'].median())
print(without_outliers['16_nasogastric_reflux_PH'].max() - without_outliers['16_nasogastric_reflux_PH'].min())

count    53.000000
mean      4.707547
std       1.982311
min       1.000000
25%       3.000000
50%       5.000000
75%       6.500000
max       7.500000
Name: 16_nasogastric_reflux_PH, dtype: float64
2.0
5.0
6.5


### Предварительные выводы по столбцам 13 - 16
1. В пункте 13 (в описании указано, что это важный параметр) данные распределились равномерно: примерно с одинаковой частотой встречается нулевое, легкое и среднее вздутие живота. Сильное вздутие наблюдается несколько реже.
2. В столбце 14 чаще всего встречаются значения "воздух не выходит из зонда" и "выходит небольшое количество воздуха". В столбце 15 самое частое значение - 1 (отсутствие назогастрального рефлюкса). Видимо, этот рефлюкс появляется только в самых тяжелых случаях.
3. Про столбец 16 интересны два факта. Во-первых, в описании указано, что значения проставляются по шкале от 0 до 14, а представлены только значения от 1 до 7,5. Во-вторых, совсем нет выбросов. Но это какая-то темная информация, столбец очень плохо заполнен, и информацию использовать было бы сложно.

## Анализ столбцов 17 - 20

In [36]:
horses_20 = horses[[headers_list[2]] + headers_list[16:20]]
horses_20.head()

3__Hospital_Number  17_rectal_examination_-_feces  18_abdomen  \
0              530101                            3.0         5.0   
1              534817                            4.0         2.0   
2              530334                            1.0         1.0   
3             5290409                            3.0         NaN   
4              530255                            NaN         NaN   

   19_packed_cell_volume  20_total_protein  
0                   45.0               8.4  
1                   50.0              85.0  
2                   33.0               6.7  
3                   48.0               7.2  
4                   74.0               7.4

In [37]:
horses_20.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 5 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   3__Hospital_Number             300 non-null    int64  
 1   17_rectal_examination_-_feces  198 non-null    float64
 2   18_abdomen                     182 non-null    float64
 3   19_packed_cell_volume          271 non-null    float64
 4   20_total_protein               267 non-null    float64
dtypes: float64(4), int64(1)
memory usage: 11.8 KB


В этой группе два столбца с категориальными показателями (17 и 18) и два столбца с количественными показателями (19 и 20).

In [38]:
horses_20.groupby(['17_rectal_examination_-_feces']).count()['3__Hospital_Number']

17_rectal_examination_-_feces
1.0    57
2.0    13
3.0    49
4.0    79
Name: 3__Hospital_Number, dtype: int64

In [39]:
horses_20.groupby(['18_abdomen']).count()['3__Hospital_Number']

18_abdomen
1.0    28
2.0    19
3.0    13
4.0    43
5.0    79
Name: 3__Hospital_Number, dtype: int64

In [40]:
# Основные метрики для столбца 19
print(horses_20['19_packed_cell_volume'].describe())
print(horses_20['19_packed_cell_volume'].round(2). mode()[0])
print(horses_20['19_packed_cell_volume'].median())
print(horses_20['19_packed_cell_volume'].max() - horses_20['19_packed_cell_volume'].min())

count    271.000000
mean      46.295203
std       10.419335
min       23.000000
25%       38.000000
50%       45.000000
75%       52.000000
max       75.000000
Name: 19_packed_cell_volume, dtype: float64
37.0
45.0
52.0


In [41]:
# Выбросы для столбца 19
q1 = horses_20['19_packed_cell_volume'].quantile(0.25)
q3 = horses_20['19_packed_cell_volume'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr)
upper_bound = q3 + (1.5 * iqr)
without_outliers = horses_20[horses_20['19_packed_cell_volume'].between(lower_bound, upper_bound)]
print(without_outliers['19_packed_cell_volume'].describe())
print(without_outliers['19_packed_cell_volume'].round(2). mode()[0])
print(without_outliers['19_packed_cell_volume'].median())
print(without_outliers['19_packed_cell_volume'].max() - without_outliers['19_packed_cell_volume'].min())

count    268.000000
mean      45.977612
std       10.031666
min       23.000000
25%       38.000000
50%       44.500000
75%       51.250000
max       73.000000
Name: 19_packed_cell_volume, dtype: float64
37.0
44.5
50.0


In [42]:
pd.concat([horses_20, without_outliers]).drop_duplicates(keep=False).dropna(subset=['19_packed_cell_volume'])

3__Hospital_Number  17_rectal_examination_-_feces  18_abdomen  \
4                530255                            NaN         NaN   
144              534519                            4.0         5.0   
151              529399                            4.0         4.0   

     19_packed_cell_volume  20_total_protein  
4                     74.0               7.4  
144                   75.0              81.0  
151                   75.0               NaN

In [43]:
# Основные метрики для столбца 20
print(horses_20['20_total_protein'].describe())
print(horses_20['20_total_protein'].round(2). mode()[0])
print(horses_20['20_total_protein'].median())
print(horses_20['20_total_protein'].max() - horses_20['20_total_protein'].min())

count    267.000000
mean      24.456929
std       27.475009
min        3.300000
25%        6.500000
50%        7.500000
75%       57.000000
max       89.000000
Name: 20_total_protein, dtype: float64
6.5
7.5
85.7


In [44]:
# Выбросы для столбца 20
q1 = horses_20['20_total_protein'].quantile(0.25)
q3 = horses_20['20_total_protein'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr)
upper_bound = q3 + (1.5 * iqr)
without_outliers = horses_20[horses_20['20_total_protein'].between(lower_bound, upper_bound)]
print(without_outliers['20_total_protein'].describe())
print(without_outliers['20_total_protein'].round(2). mode()[0])
print(without_outliers['20_total_protein'].median())
print(without_outliers['20_total_protein'].max() - without_outliers['20_total_protein'].min())

count    267.000000
mean      24.456929
std       27.475009
min        3.300000
25%        6.500000
50%        7.500000
75%       57.000000
max       89.000000
Name: 20_total_protein, dtype: float64
6.5
7.5
85.7


### Предварительные выводы по столбцам 17 - 20
1. В столбце 17 значения распределяются более-менее равномерно. Значительно реже остальных встречается значение 2 (много кала в кишечнике).
2. В столбце 18 значительно чаще остальных встречаются значения 4 и 5. Согласно описанию, это индикаторы необходимости хирургического вмешательства.
3. В столбце 19 средние значения соответствуют норме (согласно описанию, 30-50). Выбросы незначительные, встречаются только сверху. Согласно описанию, уровень эритроцитов повышается при обезвоживании или нарушении кровообращения, т. е. увеличение этого показателя является тревожным симптомом.
4. В столбце 20 нет выбросов, но очень большой разброс между средними значениями. Мода и медиана соответствуют норме, а среднее значение значительно превышает норму. Можно задуматься о стремительном увеличении этого показателя (уровень белка) при заболевании.

## Анализ столбцов 21 - 24

In [45]:
horses_24 = horses[[headers_list[2]] + headers_list[20:24]]
horses_24.head()

3__Hospital_Number  21_abdominocentesis_appearance  \
0              530101                             NaN   
1              534817                             2.0   
2              530334                             NaN   
3             5290409                             3.0   
4              530255                             NaN   

   22_abdomcentesis_total_protein  23_outcome  24_surgical_lesion?  
0                             NaN         2.0                    2  
1                             2.0         3.0                    2  
2                             NaN         1.0                    2  
3                             5.3         2.0                    1  
4                             NaN         2.0                    2

In [46]:
horses_24.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 5 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   3__Hospital_Number              300 non-null    int64  
 1   21_abdominocentesis_appearance  135 non-null    float64
 2   22_abdomcentesis_total_protein  102 non-null    float64
 3   23_outcome                      299 non-null    float64
 4   24_surgical_lesion?             300 non-null    int64  
dtypes: float64(3), int64(2)
memory usage: 11.8 KB


В этой группе только один столбец с числовыми показателями: 22 (общий белок в брюшной полости). Очень важный столбец - 23 (исход болезни), его можно будет использовать при заполнении пропусков.

In [47]:
horses_24.groupby(['21_abdominocentesis_appearance']).count()['3__Hospital_Number']

21_abdominocentesis_appearance
1.0    41
2.0    48
3.0    46
Name: 3__Hospital_Number, dtype: int64

In [48]:
# Основные метрики для столбца 22
print(horses_24['22_abdomcentesis_total_protein'].describe())
print(horses_24['22_abdomcentesis_total_protein'].round(2). mode()[0])
print(horses_24['22_abdomcentesis_total_protein'].median())
print(horses_24['22_abdomcentesis_total_protein'].max() - horses_24['22_abdomcentesis_total_protein'].min())

count    102.000000
mean       3.019608
std        1.968567
min        0.100000
25%        2.000000
50%        2.250000
75%        3.900000
max       10.100000
Name: 22_abdomcentesis_total_protein, dtype: float64
2.0
2.25
10.0


In [49]:
# Выбросы для столбца 22
q1 = horses_24['22_abdomcentesis_total_protein'].quantile(0.25)
q3 = horses_24['22_abdomcentesis_total_protein'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr)
upper_bound = q3 + (1.5 * iqr)
without_outliers = horses_24[horses_24['22_abdomcentesis_total_protein'].between(lower_bound, upper_bound)]
print(without_outliers['22_abdomcentesis_total_protein'].describe())
print(without_outliers['22_abdomcentesis_total_protein'].round(2). mode()[0])
print(without_outliers['22_abdomcentesis_total_protein'].median())
print(without_outliers['22_abdomcentesis_total_protein'].max() - without_outliers['22_abdomcentesis_total_protein'].min())

count    95.000000
mean      2.647368
std       1.413675
min       0.100000
25%       2.000000
50%       2.000000
75%       3.600000
max       6.600000
Name: 22_abdomcentesis_total_protein, dtype: float64
2.0
2.0
6.5


In [50]:
pd.concat([horses_24, without_outliers]).drop_duplicates(keep=False).dropna(subset=['22_abdomcentesis_total_protein'])

3__Hospital_Number  21_abdominocentesis_appearance  \
31               530242                             3.0   
113              527933                             NaN   
211              529893                             1.0   
225              528931                             1.0   
284              530439                             2.0   

     22_abdomcentesis_total_protein  23_outcome  24_surgical_lesion?  
31                              7.4         3.0                    1  
113                            10.0         2.0                    1  
211                            10.1         2.0                    1  
225                             8.0         1.0                    2  
284                             7.0         1.0                    2

In [51]:
horses_24.groupby(['23_outcome']).count()['3__Hospital_Number']

23_outcome
1.0    178
2.0     77
3.0     44
Name: 3__Hospital_Number, dtype: int64

In [52]:
horses_24.groupby(['24_surgical_lesion?']).count()['3__Hospital_Number']

24_surgical_lesion?
1    191
2    109
Name: 3__Hospital_Number, dtype: int64

### Предварительные выводы по столбцам 21 - 24
1. Столбцы 23 и 24 показывают что 2/3 лошадей выживают и 2/3 лошадей нужно было хирургическое вмешательство. Эвтаназия является самым редким исходом болезни (хотя один случай из шести это довольно много).
2. В столбце 21 (загазованность брюшной полости) значения распределены равномерно. Но этот столбец заполнен примерно на треть, так что выборка вряд ли репрезентативна.
3. В столбце 22(общий белок в брюшной полости) также много пропущенных данных. Выбросы идут только вверх, средние значения довольно близко к нижней части графика. В описании сказано, что чем выше этот показатель, тем хуже. Вероятно, высокие значения белка свидетельствуют о неблагоприятном течении заболевания.

## Анализ столбцов 25 - 28

In [53]:
horses_28 = horses[[headers_list[2]] + headers_list[24:28]]
horses_28.head()

3__Hospital_Number  25_type_of_lesion  26_type_of_lesion  \
0              530101              11300                  0   
1              534817               2208                  0   
2              530334                  0                  0   
3             5290409               2208                  0   
4              530255               4300                  0   

   27_type_of_lesion  28_cp_data  
0                  0           2  
1                  0           2  
2                  0           1  
3                  0           1  
4                  0           2

Эти столбцы вообще не хочется смотреть. У столбцов 25-27 описание какое-то мутное, одно на троих и почему-то много нулей. Столбец 28 в описании назван нерепрезентативным. Но можно посмотреть, какие встречаются значения и можно ли получить какую-либо интересную информацию из столбцов 25-27.

In [54]:
horses_28.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype
---  ------              --------------  -----
 0   3__Hospital_Number  300 non-null    int64
 1   25_type_of_lesion   300 non-null    int64
 2   26_type_of_lesion   300 non-null    int64
 3   27_type_of_lesion   300 non-null    int64
 4   28_cp_data          300 non-null    int64
dtypes: int64(5)
memory usage: 11.8 KB


In [55]:
horses_28.groupby(['25_type_of_lesion']).count()['3__Hospital_Number']

25_type_of_lesion
0        56
300       1
400       5
1111      1
1124      1
         ..
11400     1
12208     1
21110     1
31110     7
41110     1
Name: 3__Hospital_Number, Length: 61, dtype: int64

In [56]:
horses_28.groupby(['26_type_of_lesion']).count()['3__Hospital_Number']

26_type_of_lesion
0       293
1400      1
3111      3
3112      1
6112      1
7111      1
Name: 3__Hospital_Number, dtype: int64

In [57]:
horses_28.groupby(['27_type_of_lesion']).count()['3__Hospital_Number']

27_type_of_lesion
0       299
2209      1
Name: 3__Hospital_Number, dtype: int64

In [58]:
horses_28.groupby(['25_type_of_lesion', '26_type_of_lesion', '27_type_of_lesion']).count()['3__Hospital_Number'].head(50)

25_type_of_lesion  26_type_of_lesion  27_type_of_lesion
0                  0                  0                    56
300                0                  0                     1
400                0                  0                     5
1111               0                  0                     1
1124               0                  0                     1
1400               0                  0                     8
2111               0                  0                     3
2112               0                  0                     5
2113               0                  0                     6
2124               0                  0                     9
2205               0                  0                    13
2206               0                  0                     4
2207               0                  0                     3
2208               0                  0                    20
2209               0                  0                    10
              

In [59]:
horses_28.groupby(['25_type_of_lesion', '26_type_of_lesion', '27_type_of_lesion']).count()['3__Hospital_Number'].tail(20)

25_type_of_lesion  26_type_of_lesion  27_type_of_lesion
5400               0                  0                    4
6111               3111               0                    1
                   3112               0                    1
6112               0                  0                    2
6209               0                  0                    1
7111               0                  0                    7
7113               0                  0                    1
7209               0                  0                    3
7400               0                  0                    1
8300               0                  0                    1
8400               0                  0                    2
9000               0                  0                    1
9400               0                  0                    2
11124              0                  0                    2
11300              0                  0                    1
11400              0         

In [60]:
horses_28.groupby(['28_cp_data']).count()['3__Hospital_Number']

28_cp_data
1     99
2    201
Name: 3__Hospital_Number, dtype: int64

### Предварительные выводы по столбцам 25 - 28
1. Про столбцы 25-28 по-прежнему ничего непонятно. Предполагаю, что так могут отмечать сочетания болезней. Это единственные столбцы, где пропуски отображаются как нули. Но это видимо сделано намеренно и отражено в описании, так что ничего трогать не буду.
2. В столбце 28 преобладает второе значение (патология отсутствует).

## Пропуски

In [61]:
horses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 28 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   1__surgery?                     299 non-null    float64
 1   2__Age                          300 non-null    int64  
 2   3__Hospital_Number              300 non-null    int64  
 3   4__rectal_temperature           240 non-null    float64
 4   5__pulse                        276 non-null    float64
 5   6__respiratory_rate             242 non-null    float64
 6   7__temperature_of_extremities   244 non-null    float64
 7   8__peripheral_pulse             231 non-null    float64
 8   9__mucous_membranes             253 non-null    float64
 9   10_capillary_refill_time        268 non-null    float64
 10  11_pain                         245 non-null    float64
 11  12_peristalsis                  256 non-null    float64
 12  13_abdominal_distension         244 

1. В столбцах surgery и outcome по одному пропуску. Можно его заменить на самое популярное значение.
2. В столбцах 25-27 пропуски уже заменены константой (0). Можно так и оставить.
3. Столбец 16 слишком неинформативен. Его можно вообще удалить.
4. Категориальные столбцы заполню модой с распределением по outcome и surgical_lesion?. В этих столбцах нет пропусков, приводится информация, чем закончилась болезнь и как надо было ее лечить. Чисто интуитивно интересно было бы посмотреть, какие симптомы чаще всего наблюдаются при том или ином сочетании этих двух показателей.
5. В числовых столбцах заполню пропуски медианой, предварительно проверю различия по по outcome и surgical_lesion?.
6. В столбце 21 слишком мало данных, поэтому пропуски заполню константой. Со столбцом 22 этот номер не пройдет, потому что там данные числовые. Текстовую константу нельзя подставить, потому что изменится тип данных в столбце, а 0 нельзя подставить, потому что это отразится на вычислениях. Поэтому заполню столбец медианой по общему правилу.

In [62]:
#Сразу заполняю пропуски в столбцах 1 и 23 и избавляюсь от столбца 16
horses['1__surgery?'] = horses['1__surgery?'].fillna(1)
horses['23_outcome'] = horses['23_outcome'].fillna(1)
horses.drop(['16_nasogastric_reflux_PH'], axis=1, inplace=True)
horses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 27 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   1__surgery?                     300 non-null    float64
 1   2__Age                          300 non-null    int64  
 2   3__Hospital_Number              300 non-null    int64  
 3   4__rectal_temperature           240 non-null    float64
 4   5__pulse                        276 non-null    float64
 5   6__respiratory_rate             242 non-null    float64
 6   7__temperature_of_extremities   244 non-null    float64
 7   8__peripheral_pulse             231 non-null    float64
 8   9__mucous_membranes             253 non-null    float64
 9   10_capillary_refill_time        268 non-null    float64
 10  11_pain                         245 non-null    float64
 11  12_peristalsis                  256 non-null    float64
 12  13_abdominal_distension         244 

In [63]:
# Столбец 2: заменяю 9 на 2, чтобы значения соответствовали описанию.
horses['2__Age'].replace(9, 2, inplace=True)
horses.groupby('2__Age').count()['3__Hospital_Number']

2__Age
1    276
2     24
Name: 3__Hospital_Number, dtype: int64

In [64]:
# Столбец 4
horses.groupby(['23_outcome', '24_surgical_lesion?']).agg({'4__rectal_temperature': lambda x: x.median()})

4__rectal_temperature
23_outcome 24_surgical_lesion?                       
1.0        1                                    38.20
           2                                    38.20
2.0        1                                    38.10
           2                                    37.65
3.0        1                                    37.80
           2                                    38.30

In [65]:
# Различия незначительные, значит, пропуски заполняю медианой без детализации
horses['4__rectal_temperature'].fillna(horses['4__rectal_temperature'].median(), inplace=True)
horses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 27 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   1__surgery?                     300 non-null    float64
 1   2__Age                          300 non-null    int64  
 2   3__Hospital_Number              300 non-null    int64  
 3   4__rectal_temperature           300 non-null    float64
 4   5__pulse                        276 non-null    float64
 5   6__respiratory_rate             242 non-null    float64
 6   7__temperature_of_extremities   244 non-null    float64
 7   8__peripheral_pulse             231 non-null    float64
 8   9__mucous_membranes             253 non-null    float64
 9   10_capillary_refill_time        268 non-null    float64
 10  11_pain                         245 non-null    float64
 11  12_peristalsis                  256 non-null    float64
 12  13_abdominal_distension         244 

In [66]:
#Столбец 5
horses.groupby(['23_outcome', '24_surgical_lesion?']).agg({'5__pulse': lambda x: x.median()})

5__pulse
23_outcome 24_surgical_lesion?          
1.0        1                        64.0
           2                        48.0
2.0        1                        86.0
           2                        92.0
3.0        1                        74.0
           2                        88.0

In [67]:
# Различия значительные, значит, заполняю пропуски детализированной медианой
horses['5__pulse'].fillna(horses.groupby(['23_outcome', '24_surgical_lesion?'])['5__pulse'].transform('median'), inplace=True)
horses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 27 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   1__surgery?                     300 non-null    float64
 1   2__Age                          300 non-null    int64  
 2   3__Hospital_Number              300 non-null    int64  
 3   4__rectal_temperature           300 non-null    float64
 4   5__pulse                        300 non-null    float64
 5   6__respiratory_rate             242 non-null    float64
 6   7__temperature_of_extremities   244 non-null    float64
 7   8__peripheral_pulse             231 non-null    float64
 8   9__mucous_membranes             253 non-null    float64
 9   10_capillary_refill_time        268 non-null    float64
 10  11_pain                         245 non-null    float64
 11  12_peristalsis                  256 non-null    float64
 12  13_abdominal_distension         244 

In [68]:
# Столбец 6
horses.groupby(['23_outcome', '24_surgical_lesion?']).agg({'6__respiratory_rate': lambda x: x.median()})

6__respiratory_rate
23_outcome 24_surgical_lesion?                     
1.0        1                                   30.0
           2                                   20.0
2.0        1                                   30.0
           2                                   30.0
3.0        1                                   24.0
           2                                   20.0

In [69]:
horses['6__respiratory_rate'].fillna(horses.groupby(['23_outcome',
                                                     '24_surgical_lesion?'])['6__respiratory_rate'].transform('median'),
                                     inplace=True)
horses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 27 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   1__surgery?                     300 non-null    float64
 1   2__Age                          300 non-null    int64  
 2   3__Hospital_Number              300 non-null    int64  
 3   4__rectal_temperature           300 non-null    float64
 4   5__pulse                        300 non-null    float64
 5   6__respiratory_rate             300 non-null    float64
 6   7__temperature_of_extremities   244 non-null    float64
 7   8__peripheral_pulse             231 non-null    float64
 8   9__mucous_membranes             253 non-null    float64
 9   10_capillary_refill_time        268 non-null    float64
 10  11_pain                         245 non-null    float64
 11  12_peristalsis                  256 non-null    float64
 12  13_abdominal_distension         244 

In [70]:
# Столбец 7
horses.groupby(['23_outcome', '24_surgical_lesion?']).agg({'7__temperature_of_extremities': lambda x: x.mode()[0]})

7__temperature_of_extremities
23_outcome 24_surgical_lesion?                               
1.0        1                                              3.0
           2                                              1.0
2.0        1                                              3.0
           2                                              3.0
3.0        1                                              3.0
           2                                              3.0

In [71]:
# Значение моды меняется: при хорошем исходе и отсутствии необходимости в хирургическом вмешательстве температура
# конечностей нормальная, а в остальных случаях - сниженная. Значит, проставляю моду детализированно
horses['7__temperature_of_extremities'].fillna(horses.groupby(['23_outcome', '24_surgical_lesion?'])
                                               ['7__temperature_of_extremities'].transform(lambda x: x.mode()[0]), inplace=True)
horses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 27 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   1__surgery?                     300 non-null    float64
 1   2__Age                          300 non-null    int64  
 2   3__Hospital_Number              300 non-null    int64  
 3   4__rectal_temperature           300 non-null    float64
 4   5__pulse                        300 non-null    float64
 5   6__respiratory_rate             300 non-null    float64
 6   7__temperature_of_extremities   300 non-null    float64
 7   8__peripheral_pulse             231 non-null    float64
 8   9__mucous_membranes             253 non-null    float64
 9   10_capillary_refill_time        268 non-null    float64
 10  11_pain                         245 non-null    float64
 11  12_peristalsis                  256 non-null    float64
 12  13_abdominal_distension         244 

In [72]:
#Столбец 8
horses.groupby(['23_outcome', '24_surgical_lesion?']).agg({'8__peripheral_pulse': lambda x: x.mode()[0]})

8__peripheral_pulse
23_outcome 24_surgical_lesion?                     
1.0        1                                    1.0
           2                                    1.0
2.0        1                                    3.0
           2                                    3.0
3.0        1                                    3.0
           2                                    3.0

In [73]:
horses['8__peripheral_pulse'].fillna(horses.groupby('23_outcome')
                                               ['8__peripheral_pulse'].transform(lambda x: x.mode()[0]), inplace=True)
horses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 27 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   1__surgery?                     300 non-null    float64
 1   2__Age                          300 non-null    int64  
 2   3__Hospital_Number              300 non-null    int64  
 3   4__rectal_temperature           300 non-null    float64
 4   5__pulse                        300 non-null    float64
 5   6__respiratory_rate             300 non-null    float64
 6   7__temperature_of_extremities   300 non-null    float64
 7   8__peripheral_pulse             300 non-null    float64
 8   9__mucous_membranes             253 non-null    float64
 9   10_capillary_refill_time        268 non-null    float64
 10  11_pain                         245 non-null    float64
 11  12_peristalsis                  256 non-null    float64
 12  13_abdominal_distension         244 

In [74]:
# Столбец 9
horses.groupby(['23_outcome', '24_surgical_lesion?']).agg({'9__mucous_membranes': lambda x: x.mode()[0]})

9__mucous_membranes
23_outcome 24_surgical_lesion?                     
1.0        1                                    1.0
           2                                    1.0
2.0        1                                    3.0
           2                                    5.0
3.0        1                                    4.0
           2                                    5.0

In [75]:
horses['9__mucous_membranes'].fillna(horses.groupby(['23_outcome', '24_surgical_lesion?'])
                                               ['9__mucous_membranes'].transform(lambda x: x.mode()[0]), inplace=True)
horses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 27 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   1__surgery?                     300 non-null    float64
 1   2__Age                          300 non-null    int64  
 2   3__Hospital_Number              300 non-null    int64  
 3   4__rectal_temperature           300 non-null    float64
 4   5__pulse                        300 non-null    float64
 5   6__respiratory_rate             300 non-null    float64
 6   7__temperature_of_extremities   300 non-null    float64
 7   8__peripheral_pulse             300 non-null    float64
 8   9__mucous_membranes             300 non-null    float64
 9   10_capillary_refill_time        268 non-null    float64
 10  11_pain                         245 non-null    float64
 11  12_peristalsis                  256 non-null    float64
 12  13_abdominal_distension         244 

In [76]:
# Столбец 10
horses.groupby(['23_outcome', '24_surgical_lesion?']).agg({'10_capillary_refill_time': lambda x: x.mode()[0]})

10_capillary_refill_time
23_outcome 24_surgical_lesion?                          
1.0        1                                         1.0
           2                                         1.0
2.0        1                                         1.0
           2                                         2.0
3.0        1                                         2.0
           2                                         1.0

In [77]:
horses.groupby('10_capillary_refill_time').count()['3__Hospital_Number']

10_capillary_refill_time
1.0    188
2.0     78
3.0      2
Name: 3__Hospital_Number, dtype: int64

In [78]:
# Есть ощущение, что тройки - это человеческая ошибка (тройка как 3 секунды).
# Поэтому я меняю их не на самое популярное значение, а на 2

horses['10_capillary_refill_time'].replace(3, 2, inplace=True)
horses.groupby('10_capillary_refill_time').count()['3__Hospital_Number']

10_capillary_refill_time
1.0    188
2.0     80
Name: 3__Hospital_Number, dtype: int64

In [79]:
# Пропуски заполняю модой в разрезе двух столбцов.
horses['10_capillary_refill_time'].fillna(horses.groupby(['23_outcome', '24_surgical_lesion?'])
                                               ['10_capillary_refill_time'].transform(lambda x: x.mode()[0]), inplace=True)
horses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 27 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   1__surgery?                     300 non-null    float64
 1   2__Age                          300 non-null    int64  
 2   3__Hospital_Number              300 non-null    int64  
 3   4__rectal_temperature           300 non-null    float64
 4   5__pulse                        300 non-null    float64
 5   6__respiratory_rate             300 non-null    float64
 6   7__temperature_of_extremities   300 non-null    float64
 7   8__peripheral_pulse             300 non-null    float64
 8   9__mucous_membranes             300 non-null    float64
 9   10_capillary_refill_time        300 non-null    float64
 10  11_pain                         245 non-null    float64
 11  12_peristalsis                  256 non-null    float64
 12  13_abdominal_distension         244 

In [80]:
# Столбец 11
horses.groupby(['23_outcome', '24_surgical_lesion?']).agg({'11_pain': lambda x: x.mode()[0]})

11_pain
23_outcome 24_surgical_lesion?         
1.0        1                        3.0
           2                        1.0
2.0        1                        5.0
           2                        4.0
3.0        1                        4.0
           2                        2.0

In [81]:
horses['11_pain'].fillna(horses.groupby(['23_outcome', '24_surgical_lesion?'])
                                               ['11_pain'].transform(lambda x: x.mode()[0]), inplace=True)
horses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 27 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   1__surgery?                     300 non-null    float64
 1   2__Age                          300 non-null    int64  
 2   3__Hospital_Number              300 non-null    int64  
 3   4__rectal_temperature           300 non-null    float64
 4   5__pulse                        300 non-null    float64
 5   6__respiratory_rate             300 non-null    float64
 6   7__temperature_of_extremities   300 non-null    float64
 7   8__peripheral_pulse             300 non-null    float64
 8   9__mucous_membranes             300 non-null    float64
 9   10_capillary_refill_time        300 non-null    float64
 10  11_pain                         300 non-null    float64
 11  12_peristalsis                  256 non-null    float64
 12  13_abdominal_distension         244 

In [82]:
# Столбец 12
horses.groupby(['23_outcome', '24_surgical_lesion?']).agg({'12_peristalsis': lambda x: x.mode()[0]})

12_peristalsis
23_outcome 24_surgical_lesion?                
1.0        1                               3.0
           2                               3.0
2.0        1                               3.0
           2                               4.0
3.0        1                               4.0
           2                               3.0

In [83]:
horses['12_peristalsis'].fillna(horses.groupby(['23_outcome', '24_surgical_lesion?'])
                                               ['12_peristalsis'].transform(lambda x: x.mode()[0]), inplace=True)
horses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 27 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   1__surgery?                     300 non-null    float64
 1   2__Age                          300 non-null    int64  
 2   3__Hospital_Number              300 non-null    int64  
 3   4__rectal_temperature           300 non-null    float64
 4   5__pulse                        300 non-null    float64
 5   6__respiratory_rate             300 non-null    float64
 6   7__temperature_of_extremities   300 non-null    float64
 7   8__peripheral_pulse             300 non-null    float64
 8   9__mucous_membranes             300 non-null    float64
 9   10_capillary_refill_time        300 non-null    float64
 10  11_pain                         300 non-null    float64
 11  12_peristalsis                  300 non-null    float64
 12  13_abdominal_distension         244 

In [84]:
# Столбец 13
horses.groupby(['23_outcome', '24_surgical_lesion?']).agg({'13_abdominal_distension': lambda x: x.mode()[0]})

13_abdominal_distension
23_outcome 24_surgical_lesion?                         
1.0        1                                        1.0
           2                                        1.0
2.0        1                                        3.0
           2                                        2.0
3.0        1                                        3.0
           2                                        4.0

In [85]:
horses['13_abdominal_distension'].fillna(horses.groupby(['23_outcome', '24_surgical_lesion?'])
                                               ['13_abdominal_distension'].transform(lambda x: x.mode()[0]), inplace=True)
horses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 27 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   1__surgery?                     300 non-null    float64
 1   2__Age                          300 non-null    int64  
 2   3__Hospital_Number              300 non-null    int64  
 3   4__rectal_temperature           300 non-null    float64
 4   5__pulse                        300 non-null    float64
 5   6__respiratory_rate             300 non-null    float64
 6   7__temperature_of_extremities   300 non-null    float64
 7   8__peripheral_pulse             300 non-null    float64
 8   9__mucous_membranes             300 non-null    float64
 9   10_capillary_refill_time        300 non-null    float64
 10  11_pain                         300 non-null    float64
 11  12_peristalsis                  300 non-null    float64
 12  13_abdominal_distension         300 

In [86]:
# Столбец 14
horses.groupby(['23_outcome', '24_surgical_lesion?']).agg({'14_nasogastric_tube': lambda x: x.mode()[0]})

14_nasogastric_tube
23_outcome 24_surgical_lesion?                     
1.0        1                                    2.0
           2                                    2.0
2.0        1                                    2.0
           2                                    3.0
3.0        1                                    2.0
           2                                    2.0

In [87]:
horses['14_nasogastric_tube'].fillna(horses.groupby(['23_outcome', '24_surgical_lesion?'])
                                               ['14_nasogastric_tube'].transform(lambda x: x.mode()[0]), inplace=True)
horses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 27 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   1__surgery?                     300 non-null    float64
 1   2__Age                          300 non-null    int64  
 2   3__Hospital_Number              300 non-null    int64  
 3   4__rectal_temperature           300 non-null    float64
 4   5__pulse                        300 non-null    float64
 5   6__respiratory_rate             300 non-null    float64
 6   7__temperature_of_extremities   300 non-null    float64
 7   8__peripheral_pulse             300 non-null    float64
 8   9__mucous_membranes             300 non-null    float64
 9   10_capillary_refill_time        300 non-null    float64
 10  11_pain                         300 non-null    float64
 11  12_peristalsis                  300 non-null    float64
 12  13_abdominal_distension         300 

In [88]:
# Столбец 15
horses.groupby(['23_outcome', '24_surgical_lesion?']).agg({'15_nasogastric_reflux': lambda x: x.mode()[0]})

15_nasogastric_reflux
23_outcome 24_surgical_lesion?                       
1.0        1                                      1.0
           2                                      1.0
2.0        1                                      1.0
           2                                      2.0
3.0        1                                      1.0
           2                                      1.0

In [89]:
horses['15_nasogastric_reflux'].fillna(horses.groupby(['23_outcome', '24_surgical_lesion?'])
                                               ['15_nasogastric_reflux'].transform(lambda x: x.mode()[0]), inplace=True)
horses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 27 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   1__surgery?                     300 non-null    float64
 1   2__Age                          300 non-null    int64  
 2   3__Hospital_Number              300 non-null    int64  
 3   4__rectal_temperature           300 non-null    float64
 4   5__pulse                        300 non-null    float64
 5   6__respiratory_rate             300 non-null    float64
 6   7__temperature_of_extremities   300 non-null    float64
 7   8__peripheral_pulse             300 non-null    float64
 8   9__mucous_membranes             300 non-null    float64
 9   10_capillary_refill_time        300 non-null    float64
 10  11_pain                         300 non-null    float64
 11  12_peristalsis                  300 non-null    float64
 12  13_abdominal_distension         300 

In [90]:
# Столбец 17
horses.groupby(['23_outcome', '24_surgical_lesion?']).agg({'17_rectal_examination_-_feces': lambda x: x.mode()[0]})

17_rectal_examination_-_feces
23_outcome 24_surgical_lesion?                               
1.0        1                                              4.0
           2                                              1.0
2.0        1                                              4.0
           2                                              4.0
3.0        1                                              1.0
           2                                              4.0

In [91]:
horses['17_rectal_examination_-_feces'].fillna(horses.groupby(['23_outcome', '24_surgical_lesion?'])
                                               ['17_rectal_examination_-_feces'].transform(lambda x: x.mode()[0]), inplace=True)
horses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 27 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   1__surgery?                     300 non-null    float64
 1   2__Age                          300 non-null    int64  
 2   3__Hospital_Number              300 non-null    int64  
 3   4__rectal_temperature           300 non-null    float64
 4   5__pulse                        300 non-null    float64
 5   6__respiratory_rate             300 non-null    float64
 6   7__temperature_of_extremities   300 non-null    float64
 7   8__peripheral_pulse             300 non-null    float64
 8   9__mucous_membranes             300 non-null    float64
 9   10_capillary_refill_time        300 non-null    float64
 10  11_pain                         300 non-null    float64
 11  12_peristalsis                  300 non-null    float64
 12  13_abdominal_distension         300 

In [92]:
# Столбец 18
horses.groupby(['23_outcome', '24_surgical_lesion?']).agg({'18_abdomen': lambda x: x.mode()[0]})

18_abdomen
23_outcome 24_surgical_lesion?            
1.0        1                           5.0
           2                           1.0
2.0        1                           5.0
           2                           1.0
3.0        1                           5.0
           2                           5.0

In [93]:
horses['18_abdomen'].fillna(horses.groupby(['23_outcome', '24_surgical_lesion?'])
                                               ['18_abdomen'].transform(lambda x: x.mode()[0]), inplace=True)
horses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 27 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   1__surgery?                     300 non-null    float64
 1   2__Age                          300 non-null    int64  
 2   3__Hospital_Number              300 non-null    int64  
 3   4__rectal_temperature           300 non-null    float64
 4   5__pulse                        300 non-null    float64
 5   6__respiratory_rate             300 non-null    float64
 6   7__temperature_of_extremities   300 non-null    float64
 7   8__peripheral_pulse             300 non-null    float64
 8   9__mucous_membranes             300 non-null    float64
 9   10_capillary_refill_time        300 non-null    float64
 10  11_pain                         300 non-null    float64
 11  12_peristalsis                  300 non-null    float64
 12  13_abdominal_distension         300 

In [94]:
# Столбец 19
horses.groupby(['23_outcome', '24_surgical_lesion?']).agg({'19_packed_cell_volume': lambda x: x.median()})

19_packed_cell_volume
23_outcome 24_surgical_lesion?                       
1.0        1                                     43.0
           2                                     42.0
2.0        1                                     51.0
           2                                     64.0
3.0        1                                     50.0
           2                                     53.0

In [95]:
horses['19_packed_cell_volume'].fillna(horses.groupby(['23_outcome', '24_surgical_lesion?'])
                                               ['19_packed_cell_volume'].transform('median'), inplace=True)
horses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 27 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   1__surgery?                     300 non-null    float64
 1   2__Age                          300 non-null    int64  
 2   3__Hospital_Number              300 non-null    int64  
 3   4__rectal_temperature           300 non-null    float64
 4   5__pulse                        300 non-null    float64
 5   6__respiratory_rate             300 non-null    float64
 6   7__temperature_of_extremities   300 non-null    float64
 7   8__peripheral_pulse             300 non-null    float64
 8   9__mucous_membranes             300 non-null    float64
 9   10_capillary_refill_time        300 non-null    float64
 10  11_pain                         300 non-null    float64
 11  12_peristalsis                  300 non-null    float64
 12  13_abdominal_distension         300 

In [96]:
# Столбец 20
horses.groupby(['23_outcome', '24_surgical_lesion?']).agg({'20_total_protein': [lambda x: x.median(), 'count']})

20_total_protein      
                                     <lambda_0> count
23_outcome 24_surgical_lesion?                       
1.0        1                               7.80    85
           2                               7.30    81
2.0        1                               6.85    58
           2                               6.80     7
3.0        1                              11.60    24
           2                              63.00    12

In [97]:
# Странно большая медиана в случае 3/2. Хочу посмотреть на этот сектор подробнее
horses[(horses['23_outcome'] == 3) & (horses['24_surgical_lesion?'] == 2)]['20_total_protein']

1      85.0
36      6.0
63     82.0
97     60.0
128    61.0
144    81.0
165    64.0
179     NaN
196     7.5
205     NaN
220     NaN
276    75.0
293    62.0
295    65.0
299    62.0
Name: 20_total_protein, dtype: float64

In [98]:
# В этом секторе значения в принципе высокие за некоторыми исключениями. К тому же в столбце 20 вообще выбросов нет.
# Так что считаю данные достоверными и заполняю пропуски по общему правилу.
horses['20_total_protein'].fillna(horses.groupby(['23_outcome', '24_surgical_lesion?'])
                                               ['20_total_protein'].transform('median'), inplace=True)
horses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 27 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   1__surgery?                     300 non-null    float64
 1   2__Age                          300 non-null    int64  
 2   3__Hospital_Number              300 non-null    int64  
 3   4__rectal_temperature           300 non-null    float64
 4   5__pulse                        300 non-null    float64
 5   6__respiratory_rate             300 non-null    float64
 6   7__temperature_of_extremities   300 non-null    float64
 7   8__peripheral_pulse             300 non-null    float64
 8   9__mucous_membranes             300 non-null    float64
 9   10_capillary_refill_time        300 non-null    float64
 10  11_pain                         300 non-null    float64
 11  12_peristalsis                  300 non-null    float64
 12  13_abdominal_distension         300 

In [99]:
# В столбце 21 заполняю пустые значения константой.
horses['21_abdominocentesis_appearance'].fillna(0, inplace=True)
horses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 27 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   1__surgery?                     300 non-null    float64
 1   2__Age                          300 non-null    int64  
 2   3__Hospital_Number              300 non-null    int64  
 3   4__rectal_temperature           300 non-null    float64
 4   5__pulse                        300 non-null    float64
 5   6__respiratory_rate             300 non-null    float64
 6   7__temperature_of_extremities   300 non-null    float64
 7   8__peripheral_pulse             300 non-null    float64
 8   9__mucous_membranes             300 non-null    float64
 9   10_capillary_refill_time        300 non-null    float64
 10  11_pain                         300 non-null    float64
 11  12_peristalsis                  300 non-null    float64
 12  13_abdominal_distension         300 

In [102]:
# Столбец 22
horses.groupby(['23_outcome', '24_surgical_lesion?']).agg({'22_abdomcentesis_total_protein': [lambda x: x.median()]})

22_abdomcentesis_total_protein
                                                     <lambda>
23_outcome 24_surgical_lesion?                               
1.0        1                                              2.0
           2                                              2.3
2.0        1                                              3.4
           2                                              4.1
3.0        1                                              2.0
           2                                              2.0

In [103]:
horses['22_abdomcentesis_total_protein'].fillna(horses.groupby(['23_outcome', '24_surgical_lesion?'])
                                               ['22_abdomcentesis_total_protein'].transform('median'), inplace=True)
horses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 27 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   1__surgery?                     300 non-null    float64
 1   2__Age                          300 non-null    int64  
 2   3__Hospital_Number              300 non-null    int64  
 3   4__rectal_temperature           300 non-null    float64
 4   5__pulse                        300 non-null    float64
 5   6__respiratory_rate             300 non-null    float64
 6   7__temperature_of_extremities   300 non-null    float64
 7   8__peripheral_pulse             300 non-null    float64
 8   9__mucous_membranes             300 non-null    float64
 9   10_capillary_refill_time        300 non-null    float64
 10  11_pain                         300 non-null    float64
 11  12_peristalsis                  300 non-null    float64
 12  13_abdominal_distension         300 

Ура! Получился dataframe без пропусков! Какое счастье!